From: Benjamin Otte <otte@redhat.com>
Date: Fri, 21 Feb 2025 19:02:17 +0100
Subject: gpu: Add first node imlementations that don't blend

For certain nodes, we can glDisable(GL_BLEND) and avoid using the blend
engine. This is a tiny sppedup for fast GPUs, but I can actually measure
it to be 10% inside software rendering with valgrind.

Hopefully it's a large speedup for older devices.

Another benefit is that we can avoid clearing the background, because
we don't blend with it, so using LOAD_OP_DONT_CARE is fine.

Related: #7146
---
 gsk/gpu/gskgpunodeprocessor.c | 40 ++++++++++++++++++++++++++++++++--------
 1 file changed, 32 insertions(+), 8 deletions(-)

diff --git a/gsk/gpu/gskgpunodeprocessor.c b/gsk/gpu/gskgpunodeprocessor.c
index 2c1e930..8ed5ec5 100644
--- a/gsk/gpu/gskgpunodeprocessor.c
+++ b/gsk/gpu/gskgpunodeprocessor.c
@@ -2200,6 +2200,30 @@ gsk_gpu_node_processor_add_texture_node (GskGpuNodeProcessor *self,
   g_object_unref (image);
 }
 
+static gboolean
+gsk_gpu_node_processor_add_first_node_no_blend (GskGpuNodeProcessor *self,
+                                                GskGpuFirstNodeInfo *info,
+                                                GskRenderNode       *node)
+{
+  if (!node->fully_opaque)
+    return FALSE;
+
+  if (!gsk_gpu_node_processor_clip_first_node (self, info, &node->bounds))
+    return FALSE;
+
+  gsk_gpu_first_node_begin_rendering (self, info, NULL);
+
+  self->blend = GSK_GPU_BLEND_NONE;
+  self->pending_globals |= GSK_GPU_GLOBAL_BLEND;
+
+  gsk_gpu_node_processor_add_node (self, node);
+
+  self->blend = GSK_GPU_BLEND_OVER;
+  self->pending_globals |= GSK_GPU_GLOBAL_BLEND;
+
+  return TRUE;
+}
+
 static GskGpuImage *
 gsk_gpu_get_texture_node_as_image (GskGpuFrame           *frame,
                                    GskGpuAsImageFlags     flags,
@@ -3872,7 +3896,7 @@ static const struct
     0,
     GSK_GPU_HANDLE_OPACITY,
     gsk_gpu_node_processor_add_cairo_node,
-    NULL,
+    gsk_gpu_node_processor_add_first_node_no_blend,
     gsk_gpu_get_cairo_node_as_image,
   },
   [GSK_COLOR_NODE] = {
@@ -3886,35 +3910,35 @@ static const struct
     0,
     GSK_GPU_HANDLE_OPACITY,
     gsk_gpu_node_processor_add_linear_gradient_node,
-    NULL,
+    gsk_gpu_node_processor_add_first_node_no_blend,
     NULL,
   },
   [GSK_REPEATING_LINEAR_GRADIENT_NODE] = {
     0,
     GSK_GPU_HANDLE_OPACITY,
     gsk_gpu_node_processor_add_linear_gradient_node,
-    NULL,
+    gsk_gpu_node_processor_add_first_node_no_blend,
     NULL,
   },
   [GSK_RADIAL_GRADIENT_NODE] = {
     0,
     GSK_GPU_HANDLE_OPACITY,
     gsk_gpu_node_processor_add_radial_gradient_node,
-    NULL,
+    gsk_gpu_node_processor_add_first_node_no_blend,
     NULL,
   },
   [GSK_REPEATING_RADIAL_GRADIENT_NODE] = {
     0,
     GSK_GPU_HANDLE_OPACITY,
     gsk_gpu_node_processor_add_radial_gradient_node,
-    NULL,
+    gsk_gpu_node_processor_add_first_node_no_blend,
     NULL,
   },
   [GSK_CONIC_GRADIENT_NODE] = {
     0,
     GSK_GPU_HANDLE_OPACITY,
     gsk_gpu_node_processor_add_conic_gradient_node,
-    NULL,
+    gsk_gpu_node_processor_add_first_node_no_blend,
     NULL,
   },
   [GSK_BORDER_NODE] = {
@@ -3928,7 +3952,7 @@ static const struct
     0,
     GSK_GPU_HANDLE_OPACITY,
     gsk_gpu_node_processor_add_texture_node,
-    NULL,
+    gsk_gpu_node_processor_add_first_node_no_blend,
     gsk_gpu_get_texture_node_as_image,
   },
   [GSK_INSET_SHADOW_NODE] = {
@@ -4040,7 +4064,7 @@ static const struct
     0,
     0,
     gsk_gpu_node_processor_add_texture_scale_node,
-    NULL,
+    gsk_gpu_node_processor_add_first_node_no_blend,
     NULL,
   },
   [GSK_MASK_NODE] = {
